src/runtime/refiner.ts

```diff
@@ -956,3 +956,138 @@ export class SQLExtracter {
     return `SELECT * FROM ${table}` + (filterTerms.length ? ` WHERE ${filterTerms.join(' AND ')}` : '') + ';';
   }
 }
+
+export class Fraction {
+  num: Polynomial;
+  den: Polynomial;
+
+  constructor(n?: Polynomial, d?: Polynomial) {
+    this.num = n ? Polynomial.copyOf(n) : new Polynomial();
+    this.den = d ? Polynomial.copyOf(d) : new Polynomial([1]);
+    if (this.den.isZero()) {
+      throw new Error('Division by zero.');
+    }
+    this.reduce();
+  }
+
+  static add(a: Fraction, b: Fraction): Fraction {
+    const den = Polynomial.multiply(a.den, b.den);
+    const num = Polynomial.add(Polynomial.multiply(a.num, b.den), Polynomial.multiply(b.num, a.den));
+    return new Fraction(num, den);
+  }
+
+  static negate(a: Fraction): Fraction {
+    return new Fraction(Polynomial.negate(a.num), a.den);
+  }
+
+  static subtract(a: Fraction, b: Fraction): Fraction {
+    const negB = Fraction.negate(b);
+    return Fraction.add(a, negB);
+  }
+
+  static multiply(a: Fraction, b: Fraction): Fraction {
+    return new Fraction(Polynomial.multiply(a.num, b.num), Polynomial.multiply(a.den, b.den));
+  }
+
+  static divide(a: Fraction, b: Fraction): Fraction {
+    const invB = new Fraction(b.den, b.num);
+    return Fraction.multiply(a, invB);
+  }
+
+  reduce() {
+    if (this.num.isZero()) {
+      this.den = new Polynomial([1]);
+      return;
+    }
+    if (this.num.isConstant() && this.den.isConstant()) {
+      this.num = new Polynomial([this.num.coeffs[0]/this.den.coeffs[0]]);
+      this.den = new Polynomial([1]);
+      return;
+    }
+    // TODO(ragdev): Fractions can be reduced further by factoring out the gcd of
+    // the coeffs in num and den, and then dividing the two. However, since the numbers are floating
+    // points, the precision and computation cost of gcd function will be a trade-off to consider.
+  }
+}
+
+export class Polynomial {
+  _coeffs: number[];
+
+  constructor(coeffs?: number[]) {
+    if (coeffs) {
+      this.coeffs = coeffs;
+    } else {
+      this.coeffs = [0];
+    }
+  }
+
+  static copyOf(pn: Polynomial): Polynomial {
+    const copy = new Polynomial();
+    copy.coeffs = pn.coeffs;
+    return copy;
+  }
+
+  get coeffs(): number[] {
+    for (let i = this._coeffs.length - 1; i > 0; i -= 1) {
+      if (this._coeffs[i] === 0) {
+        this._coeffs.splice(i, 1);
+      } else {
+        break;
+      }
+  }
```

Comment:
Possibly equivalent to
```
while(this._coeffs.length > 0 && this._coeffs[0] == 0) {
  this._coeffs.pop();
}
```

Reply:
I think it's equivalent to:
```
while(this._coeffs.length >= 1 && this._coeffs[this._coeffs.length - 1] === 0) {
      this._coeffs.pop();
    }
```
But using a while instead of for here is a lot neater. Thanks!

Reply:
Yup, that looks better.